### PR TITLE
IBM Support 

### PR DESCRIPTION
**Pull request by sarahwooders**
Draft pull request to track remaining todos with IBM integration. 

TODOs: 
- [x] Pull/merge recent skyplane changes 
- [x] Test multipart transfers 
- [x] Query bucket locations rather than reading location from config
- [x] If possible, automate generation of ibm_credentials file as part of `skyplane init` (using https://cloud.ibm.com/docs/cloud-object-storage?topic=cloud-object-storage-uhc-hmac-credentials-main) 
- [x] Update documentation to include IBM setup details and transfer examples

### CODE REVIEW

**Comment by gilv** [comment]
@sarahwooders I already support to query bucket location. Unfortunately, IBM Cloud Object Storage doesn't has a single API to query a location, so the way is to pass over all regions and check if bucket exists there. https://github.com/skyplane-project/skyplane/blob/ibmcloud-support/skyplane/obj_store/cos_interface.py#L37.
Does this address the bullet you asked: Query bucket locations rather than reading location from config

**Comment by gilv** [comment]
@sarahwooders also all uploads done with multipart. What do you mean by "Test multipart transfers" ?

**Comment by gilv** [comment]
@sarahwooders I follow https://skyplane.org/en/latest/contributing.html and execute `black -l 140 .   ` but I see that checks failed in PR on the black type?

**Comment by sarahwooders** [comment]
> @sarahwooders I follow https://skyplane.org/en/latest/contributing.html and execute `black -l 140 . ` but I see that checks failed in PR on the black type? This is in my laptop:
> 
> ```
> `MacBook-Pro skyplane % black . --check --line-length=140 
> All done! ✨ 🍰 ✨
> 117 files would be left unchanged.
> MacBook-Pro skyplane % git status
> On branch ibmcloud-support
> Your branch is up to date with 'skyplane/ibmcloud-support'.
> 
> nothing to commit, working tree clean
> 
> `
> ```
> 
> but the pytest / black-pytype (pull_request) Failing after 14m fails, Can you check please?

I think there are different versions of black - could you please try using poetry instead and run `poetry run black . --line-length=140 && poetry run pytype . --config .pytype.cfg -j 16`?

**Comment by gilv** [comment]
@sarahwooders this also returned the same as before

    `kyplane % poetry run black . --line-length=140
    Creating virtualenv skyplane-4B0aA3OF-py3.9 in /Users/gilvernik/Library/Caches/pypoetry/virtualenvs
    All done! ✨ 🍰 ✨
    117 files left unchanged.
    `
If you run the same, does it show you different things?

**Comment by sarahwooders** [comment]
Did you run `pip install -r requirements-dev.txt`? I ran the formatter and got some changes, so just pushed them.

**Comment by sarahwooders** [comment]
Also one issue that I've continued to have is that `ibmcloud_resource_group_id` doesn't get automatically added to `~/.skyplane/config` (I have to manually copy it from `~/.bluemix/ibm_credentials`). I think you can reproduce this by removing `~/.skyplane` and then re-running `skyplane init`.

**Comment by gilv** [comment]
@sarahwooders black works now...thanks. 
I see that  pytest / black-pytype still fails. Where this module suppose to be installed?

  ```
  /home/runner/.virtualenvs/skyplane-sLdV35Eb-py3.10/lib/python3.10/site-packages/urllib3/contrib/socks.py:50: DependencyWarning: SOCKS support in urllib3 requires the installation of optional dependencies: specifically, PySocks.  For more information, see https://urllib3.readthedocs.io/en/1.26.x/contrib.html#socks-proxies
  [13](https://github.com/skyplane-project/skyplane/actions/runs/4445310069/jobs/7804307955?pr=731#step:8:14)
    warnings.warn(
  [14](https://github.com/skyplane-project/skyplane/actions/runs/4445310069/jobs/7804307955?pr=731#step:8:15)
  ninja: Entering directory `.pytype'
  [15](https://github.com/skyplane-project/skyplane/actions/runs/4445310069/jobs/7804307955?pr=731#step:8:16)
  [1/122] check skyplane.gateway_version
  [16](https://github.com/skyplane-project/skyplane/actions/runs/4445310069/jobs/7804307955?pr=731#step:8:17)
  [2/122] check skyplane.exceptions
  [17](https://github.com/skyplane-project/skyplane/actions/runs/4445310069/jobs/7804307955?pr=731#step:8:18)
  [3/122] check skyplane.utils.definitions
  [18](https://github.com/skyplane-project/skyplane/actions/runs/4445310069/jobs/7804307955?pr=731#step:8:19)
  [4/122] check skyplane.config
  [19](https://github.com/skyplane-project/skyplane/actions/runs/4445310069/jobs/7804307955?pr=731#step:8:20)
  [5/122] check skyplane.utils.logger
  [20](https://github.com/skyplane-project/skyplane/actions/runs/4445310069/jobs/7804307955?pr=731#step:8:21)
  [6/122] check skyplane.utils.imports
  [21](https://github.com/skyplane-project/skyplane/actions/runs/4445310069/jobs/7804307955?pr=731#step:8:22)
  [7/122] check skyplane.config_paths
  [22](https://github.com/skyplane-project/skyplane/actions/runs/4445310069/jobs/7804307955?pr=731#step:8:23)
  [8/122] check skyplane.utils.timer
  [23](https://github.com/skyplane-project/skyplane/actions/runs/4445310069/jobs/7804307955?pr=731#step:8:24)
  [9/122] check skyplane.compute.ibmcloud.ibm_gen2.config
  [24](https://github.com/skyplane-project/skyplane/actions/runs/4445310069/jobs/7804307955?pr=731#step:8:25)
  [10/122] check skyplane.compute.ibmcloud.ibmcloud_auth
  [25](https://github.com/skyplane-project/skyplane/actions/runs/4445310069/jobs/7804307955?pr=731#step:8:26)
  FAILED: /home/runner/work/skyplane/skyplane/.pytype/pyi/skyplane/compute/ibmcloud/ibmcloud_auth.pyi 
  [26](https://github.com/skyplane-project/skyplane/actions/runs/4445310069/jobs/7804307955?pr=731#step:8:27)
  /home/runner/.virtualenvs/skyplane-sLdV35Eb-py3.10/bin/python -m pytype.single --imports_info /home/runner/work/skyplane/skyplane/.pytype/imports/skyplane.compute.ibmcloud.ibmcloud_auth.imports --module-name skyplane.compute.ibmcloud.ibmcloud_auth --platform linux -V 3.10 -o /home/runner/work/skyplane/skyplane/.pytype/pyi/skyplane/compute/ibmcloud/ibmcloud_auth.pyi --analyze-annotated --nofail --quick /home/runner/work/skyplane/skyplane/skyplane/compute/ibmcloud/ibmcloud_auth.py
  [27](https://github.com/skyplane-project/skyplane/actions/runs/4445310069/jobs/7804307955?pr=731#step:8:28)
  File "/home/runner/work/skyplane/skyplane/skyplane/compute/ibmcloud/ibmcloud_auth.py", line 34, in get_iam_authenticator: Can't find module 'ibm_cloud_sdk_core.authenticators'. [import-error]
  [28](https://github.com/skyplane-project/skyplane/actions/runs/4445310069/jobs/7804307955?pr=731#step:8:29)
  
  [30](https://github.com/skyplane-project/skyplane/actions/runs/4445310069/jobs/7804307955?pr=731#step:8:31)
  For more details, see https://google.github.io/pytype/errors.html#import-error
  [31](https://github.com/skyplane-project/skyplane/actions/runs/4445310069/jobs/7804307955?pr=731#step:8:32)
  [11/122] check skyplane.utils.generator
  [32](https://github.com/skyplane-project/skyplane/actions/runs/4445310069/jobs/7804307955?pr=731#step:8:33)
  ninja: build stopped: subcommand failed.
  [33](https://github.com/skyplane-project/skyplane/actions/runs/4445310069/jobs/7804307955?pr=731#step:8:34)
  Computing dependencies
  [34](https://github.com/skyplane-project/skyplane/actions/runs/4445310069/jobs/7804307955?pr=731#step:8:35)
  Analyzing 94 sources with 2 local dependencies
  [35](https://github.com/skyplane-project/skyplane/actions/runs/4445310069/jobs/7804307955?pr=731#step:8:36)
  Leaving directory '.pytype'
  [35](https://github.com/skyplane-project/skyplane/actions/runs/4445310069/jobs/7804307955?pr=731#step:8:36)
  Error: Process completed with exit code 1.
  ```

**Comment by gilv** [comment]
@sarahwooders I deleted the `~/.skyplane` folder. Then executed `skyplane init` and `resource_id` was created in the config file. I can't reproduce a flow where this filed is not created in the config file

**Comment by gilv** [comment]
@sarahwooders related your comment : If possible, automate generation of ibm_credentials file as part of skyplane init.
I didn't used ibmcloud tool as it's confusing when need to setup both VPC and Cloud Object Storage. Unfortunately this is how this tool works. The less confusing approach is just to setup all in the UI of IBM Cloud. I also assumed that those who use SkyPlane with IBM Cloud already has IBM Cloud account and therefore they have their credentials to the cloud object storage. I think it's not a good idea if SkyPlane will auto-generate credentials for COS, since it may confuse those who already has active COS account and those who doesn't  - still better create all in UI and not via CLI tool

File: skyplane/api/dataplane.py [hunk 1]
```diff
@@ -211,6 +212,7 @@ def deprovision(self, max_jobs: int = 64, spinner: bool = False):
         :param spinner: Whether to show the spinner during the job (default: False)
         :type spinner: bool
         """
+        debug = True
```
Comment (simon-mo) [inline skyplane/api/dataplane.py:215]:
```suggestion
```

File: skyplane/cli/cli.py [hunk 2]
```diff
@@ -90,4 +91,7 @@ def ssh():
 typer_click_object = typer.main.get_command(app)
 
 if __name__ == "__main__":
+    import os
+
+    os.environ["SKYPLANE_DOCKER_IMAGE"] = "public.ecr.aws/s6m1p0n8/skyplane-ibm:ibmcloud-support"
```
Comment (simon-mo) [inline skyplane/cli/cli.py:96]:
```suggestion
```

File: skyplane/cli/cli_transfer.py [hunk 3]
```diff
@@ -54,25 +54,40 @@ class SkyplaneCLI:
...
+            aws_config=self.aws_config,
+            azure_config=self.azure_config,
+            gcp_config=self.gcp_config,
+            transfer_config=self.transfer_config,
+            ibmcloud_config=self.ibmcloud_config,
         )
         typer.secho(f"Using Skyplane version {skyplane.__version__}", fg="bright_black")
         typer.secho(f"Logging to: {self.client.log_dir / 'client.log'}", fg="bright_black")
 
+        # import pdb; pdb.set_trace()
```
Comment (simon-mo) [inline skyplane/cli/cli_transfer.py:69]:
```suggestion
```

**Review by simon-mo — APPROVED** [review]
After these clean up, I think this PR is good to merge!

**Pull request merged**